org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java

```diff
@@ -55,8 +55,6 @@ public FlowInfo analyseCode(BlockScope currentScope, FlowContext flowContext, Fl
 
 	flowInfo = this.expression.analyseCode(currentScope, flowContext, flowInfo);
 	this.expression.checkNPEbyUnboxing(currentScope, flowContext, flowInfo);
-	if (flowInfo.reachMode() == FlowInfo.REACHABLE && currentScope.compilerOptions().isAnnotationBasedNullAnalysisEnabled)
-		checkAgainstNullAnnotation(currentScope, flowContext, flowInfo, this.expression);
```

Comment:
So you successfully removed a bogus analysis for the price that NPE caused by `yield` will not be detected at all, right?

I admit that the current code is only useful in the narrow situation of `return switch(v) { ... yield x; }`

I believe the better course of action would be to identify the expected annotated type to which the yield value will be assigned, without taking the shortcut to the method return type.

Reply:
> So you successfully removed a bogus analysis for the price that NPE caused by `yield` will not be detected at all, right?
> 
> I admit that the current code is only useful in the narrow situation of `return switch(v) { ... yield x; }`
>

Hold on, please take a look again! 

I added a new test via a follow up commit that proves that a null injected by `yield` of a `switch` expression that is the directly `return`ed *will be* detected: See `org.eclipse.jdt.core.tests.compiler.regression.NullAnnotationTests21.testIssue2522_2()`

Why would yield require special treatment under return anymore than ternary expression would ? 
 
> I believe the better course of action would be to identify the expected annotated type to which the yield value will be assigned, without taking the shortcut to the method return type.

This happens automatically! Just the same way this is complained against:

```
@NonNull Stuff l = switch (pm) {
					case Stuff s -> s;
					case null -> null; // error: Null type mismatch: required 'PatternMatching.@NonNull Stuff' but the provided value is null
					};
```

So do you have a snippet that shows a case where we would not detect a null injection due to the change here ?? I

Reply:
Apologies, I forgot about how a switch expression collects result expressions and how various null-checking methods will drill into those result expressions, without the yield expression needing to do anything. The fact that I wrote my previous review comment on a train is only a weak excuse.